posts/2023-07-24_rounding/rounding.qmd

```diff
@@ -0,0 +1,142 @@
+
+-   R documentation: Base R [Round](https://www.rdocumentation.org/packages/base/versions/3.6.2/topics/Round), [`janitor::round_half_up()`](https://sfirke.github.io/janitor/reference/round_half_up.html), [`tidytlg::roundSAS()`](https://pharmaverse.github.io/tidytlg/main/reference/roundSAS.html)
+
+-   [CAMIS](https://psiaims.github.io/CAMIS/) (Comparing Analysis Method Implementations in Software)
+
+## Round half up in R
+
+The motivation for having a 'round half up' function is clear: it's a widely used rounding method, but there are no such options available in base R.
+
+There are multiple forums that discussed this topic, and a few functions available in some R packages.
+Hope you won't need to spend much time identifying a safe option for rounding half up in R after reading the following section.
+
+### Bug fix
+
+The first time I needed to round half up in R, I chose the function from a [PHUSE paper](https://www.lexjansen.com/phuse-us/2020/ct/CT05.pdf) and applied it to my study.
+It works fine most of the time, but has issues like the following example -
+
+```{r, message = FALSE}
+# a function that rounds half up
+# exact copy from: https://www.lexjansen.com/phuse-us/2020/ct/CT05.pdf
+ut_round <- function(x, n=0)
+{
+ # x is the value to be rounded
+ # n is the precision of the rounding
+ scale <- 10^n
+ y <- trunc(x * scale + sign(x) * 0.5) / scale
+ # Return the rounded number
+ return(y)
+}
+# round half up for 2436.845, with 2 decimal places
+ut_round(2436.845, 2)
+```
+
+The expected result is 2436.85, but the output rounds it down.
+The function's logic seems fine, I was not able to find the root cause.
+Thanks to the community effort, there are already discussions and resolution available in a [StackOverflow post](https://stackoverflow.com/questions/12688717/round-up-from-5#comment110611119_12688836) -
+
+> There are numerical precision issues, e.g., `round2(2436.845, 2)` returns `2436.84.` Changing `z + 0.5 to z + 0.5 + sqrt(.Machine$double.eps)` seems to work for me. -- Gregor Thomas Jun 24, 2020 at 2:16
+
+After fixing this bug:
+
+```{r, message = FALSE}
+# revised rounds half up
+ut_round1 <- function(x, n=0)
+{
+ # x is the value to be rounded
+ # n is the precision of the rounding
+ scale <- 10^n
+ y <- trunc(x * scale + sign(x) * 0.5 + sqrt(.Machine$double.eps)) / scale
+ # Return the rounded number
+ return(y)
+}
+# round half up for 2436.845, with 2 decimal places
+ut_round1(2436.845, 2)
+```
+
+### We are not alone
+
+The same issue occurred in the following ones as well, and has been raised by users:
+
+-   `janitor::round_half_up()`: [issue](https://github.com/sfirke/janitor/issues/396) was raised and fixed in v2.1.0
+
+-   `Tplyr`: `options(tplyr.IBMRounding = TRUE)`, [issue](https://github.com/atorus-research/Tplyr/issues/124) was raised
+
+-   `scrutiny::round_up_from()/round_up()`: [issue](https://lhdjung.github.io/scrutiny/reference/index.html#rounding) was raised
+
+-   ...
```

Comment:
replace "..." with something like "... and many others!" as for a second i thought you'd forgotten to fill something in!